File: ios/Sources/CompanionCore/BrowserLiveClient.swift
import Foundation

/// What arrives on `GET /api/bots/:id/browser/live`.
///
/// The server normalises every message before it reaches us
/// (`normalizeBrowserLiveMessage`), so these are the only shapes possible and
/// an unknown one is a protocol change, not untrusted input.
public enum BrowserLiveMessage: Sendable, Equatable {
    case frame(BrowserFrame)
    case status(BrowserStatus)
    case url(String)
    case tabs([BrowserTab])
    case ready(viewerId: String)
    case control(controlling: Bool, held: Bool)
    case heartbeat
    case error(String)
}

public struct BrowserFrame: Sendable, Equatable {
    public let seq: Int
    /// Base64 JPEG or PNG, exactly as the server framed it.
    public let data: String
    public let format: String
    public let deviceWidth: Double
    public let deviceHeight: Double

    public init(seq: Int, data: String, format: String, deviceWidth: Double, deviceHeight: Double) {
        self.seq = seq
        self.data = data
        self.format = format
        self.deviceWidth = deviceWidth
        self.deviceHeight = deviceHeight
    }

    /// Decoded bytes, or nil when the base64 was not what it claimed to be.
    /// Returning nil rather than throwing keeps the caller a view.
    public var bytes: Data? { Data(base64Encoded: data) }
}

public struct BrowserStatus: Sendable, Equatable {
    public let connected: Bool
    public let screencasting: Bool
    public let viewportWidth: Double
    public let viewportHeight: Double

    public init(connected: Bool, screencasting: Bool, viewportWidth: Double, viewportHeight: Double) {
        self.connected = connected
        self.screencasting = screencasting
        self.viewportWidth = viewportWidth
        self.viewportHeight = viewportHeight
    }
}

public struct BrowserTab: Sendable, Equatable, Identifiable {
    public let tabId: String
    public let title: String
    public let url: String
    public let active: Bool

    public init(tabId: String, title: String, url: String, active: Bool) {
        self.tabId = tabId
        self.title = title
        self.url = url
        self.active = active
    }

    public var id: String { tabId }
}

/// Decodes one server message.
///
/// Written by hand because the wire is a tagged union keyed on `type`, and
/// because a message we do not understand must be dropped rather than crash a
/// stream the person is watching.
public enum BrowserLiveDecoder {
    /// The message type travels in the SSE `event:` name, not in the payload —
    /// the server strips it (`const { type, ...data } = message`) before
    /// writing the frame. Keying on the payload instead decoded nothing at
    /// all, which is exactly how this was found.
    public static func message(event: String?, data json: String) -> BrowserLiveMessage? {
        guard let event else { return nil }
        let object = (json.data(using: .utf8).flatMap { try? JSONSerialization.jsonObject(with: $0) }) as? [String: Any] ?? [:]

        switch event {
        case "frame":
            guard let seq = object["seq"] as? Int,
                  let payload = object["data"] as? String,
                  let metadata = object["metadata"] as? [String: Any],
                  let width = number(metadata["deviceWidth"]),
                  let height = number(metadata["deviceHeight"])
            else { return nil }
            return .frame(BrowserFrame(
                seq: seq,
                data: payload,
                format: (object["format"] as? String) ?? "jpeg",
                deviceWidth: width,
                deviceHeight: height
            ))

        case "status":
            guard let connected = object["connected"] as? Bool else { return nil }
            return .status(BrowserStatus(
                connected: connected,
                screencasting: (object["screencasting"] as? Bool) ?? false,
                viewportWidth: number(object["viewportWidth"]) ?? 1280,
                viewportHeight: number(object["viewportHeight"]) ?? 720
            ))

        case "url":
            guard let url = object["url"] as? String else { return nil }
            return .url(url)

        case "tabs":
            guard let raw = object["tabs"] as? [[String: Any]] else { return nil }
            return .tabs(raw.compactMap { tab in
                guard let id = tab["tabId"] as? String else { return nil }
                return BrowserTab(
                    tabId: id,
                    title: (tab["title"] as? String) ?? "",
                    url: (tab["url"] as? String) ?? "",
                    active: (tab["active"] as? Bool) ?? false
                )
            })

        // The server names this `ready`, and it is the only place a viewer id
        // ever arrives. Without it there is nothing to post an action against.
        case "ready":
            guard let id = object["viewerId"] as? String else { return nil }
            return .ready(viewerId: id)

        case "control":
            return .control(
                controlling: (object["controlling"] as? Bool) ?? false,
                held: (object["held"] as? Bool) ?? false
            )

        case "heartbeat":
            return .heartbeat

        case "error":
            return .error((object["message"] as? String) ?? "The browser stream was interrupted.")

        default:
            // A name we do not know is a protocol change, not a reason to
            // tear down a stream the person is watching.
            return nil
        }
    }

    /// JSONSerialization hands back Int or Double depending on the literal, so
    /// a width of `1280` and one of `1280.0` must both survive.
    private static func number(_ value: Any?) -> Double? {
        (value as? Double) ?? (value as? Int).map(Double.init)
    }
}

/// The browser-live transport: one SSE stream in, one action channel out.
public struct BrowserLiveClient: Sendable {
    private let connection: Connection
    private let token: String?
    private let session: URLSession

    public init(connection: Connection, token: String?, session: URLSession = .shared) {
        self.connection = connection
        self.token = token
        self.session = session
    }

    private func request(_ method: String, _ path: String, body: Data? = nil) throws -> URLRequest {
        guard let base = connection.baseURL,
              var components = URLComponents(url: base, resolvingAgainstBaseURL: false)
        else { throw APIError.badURL }
        components.path = path
        guard let url = components.url else { throw APIError.badURL }

        var request = URLRequest(url: url)
        request.httpMethod = method
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = body
        }
        return request
    }

    /// The frame stream. Runs until the server ends it or the consuming task
    /// is cancelled; reconnection belongs to whatever knows if the view is
    /// still on screen, exactly as it does for the main event stream.
    public func live(botId: String) -> AsyncThrowingStream<BrowserLiveMessage, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = try request("GET", "/api/bots/\(botId)/browser/live")
                    // A frame stream has no business timing out: it is idle
                    // whenever the page is.
                    request.timeoutInterval = .infinity
                    let (bytes, response) = try await session.bytes(for: request)
                    if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                        // Read the refusal rather than discard it. Throwing a
                        // bare status made every 403 read as "browser control
                        // is off", whatever the server actually said.
                        var body = Data()
                        for try await byte in bytes {
                            body.append(byte)
                            if body.count >= 4096 { break }
                        }
                        throw APIError.status(code: http.statusCode, message: BrowserLiveClient.errorText(body))
                    }

                    var parser = SSEParser()
                    var line = [UInt8]()
                    // A byte at a time, and deliberately not `bytes.lines`:
                    // that folds consecutive newlines together, and a blank
                    // line is exactly what ends an SSE event. See SSE.swift.
                    for try await byte in bytes {
                        if byte == UInt8(ascii: "\n") {
                            let text = String(decoding: line, as: UTF8.self)
                            line.removeAll(keepingCapacity: true)
                            if let event = parser.line(text),
                               let message = BrowserLiveDecoder.message(event: event.event, data: event.data) {
                                continuation.yield(message)
                            }
                        } else {
                            line.append(byte)
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// One action. The server answers only after the browser has applied it,
    /// which is what makes the queue's one-in-flight rule necessary.
    @discardableResult
    public func action(botId: String, viewerId: String, body: [String: Any]) async throws -> [String: Any] {
        var payload = body
        payload["viewerId"] = viewerId
        let encoded = try JSONSerialization.data(withJSONObject: payload)
        let request = try request("POST", "/api/bots/\(botId)/browser/action", body: encoded)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.badURL }
        let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        guard http.statusCode == 200 else {
            throw APIError.status(code: http.statusCode, message: object?["error"] as? String)
        }
        return object ?? [:]
    }

    /// The `error` field of a JSON refusal, which is how both the sidecar and
    /// the harness explain themselves.
    static func errorText(_ data: Data) -> String? {
        (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
    }

    /// An input body, posted through the same channel.
    public func send(botId: String, viewerId: String, input: BrowserInputBody) async throws {
        let encoded = try JSONEncoder().encode(input)
        guard var object = (try? JSONSerialization.jsonObject(with: encoded)) as? [String: Any] else {
            throw APIError.badURL
        }
        object["viewerId"] = viewerId
        _ = try await action(botId: botId, viewerId: viewerId, body: object)
    }
}

File: android/core/src/main/kotlin/com/openmausbot/companion/core/BrowserLiveClient.kt
package com.openmausbot.companion.core

import java.io.IOException
import java.util.concurrent.atomic.AtomicReference
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * What arrives on `GET /api/bots/:id/browser/live`.
 *
 * The server normalises every message before it reaches us
 * (`normalizeBrowserLiveMessage`), so these are the only shapes possible and
 * an unknown one is a protocol change, not untrusted input.
 */
sealed interface BrowserLiveMessage {
    data class Frame(val frame: BrowserFrame) : BrowserLiveMessage
    data class Status(val status: BrowserStatus) : BrowserLiveMessage
    data class Url(val url: String) : BrowserLiveMessage
    data class Tabs(val tabs: List<BrowserTab>) : BrowserLiveMessage
    data class Ready(val viewerId: String) : BrowserLiveMessage
    data class Control(val controlling: Boolean, val held: Boolean) : BrowserLiveMessage
    data object Heartbeat : BrowserLiveMessage
    data class Error(val message: String) : BrowserLiveMessage
}

data class BrowserFrame(
    val seq: Int,
    /** Base64 JPEG or PNG, exactly as the server framed it. */
    val data: String,
    val format: String,
    val deviceWidth: Double,
    val deviceHeight: Double,
)

data class BrowserStatus(
    val connected: Boolean,
    val screencasting: Boolean,
    val viewportWidth: Double,
    val viewportHeight: Double,
)

data class BrowserTab(
    val tabId: String,
    val title: String,
    val url: String,
    val active: Boolean,
)

/**
 * Decodes one server message.
 *
 * Written by hand because the wire is a tagged union keyed on `type`, and
 * because a message we do not understand must be dropped rather than tear
 * down a stream the person is watching.
 *
 * Mirrors CompanionCore's `BrowserLiveDecoder`.
 */
object BrowserLiveDecoder {
    private val json = Json { ignoreUnknownKeys = true }

    private fun JsonObject.string(key: String) = this[key]?.jsonPrimitive?.contentOrNull
    private fun JsonObject.bool(key: String) = this[key]?.jsonPrimitive?.booleanOrNull
    private fun JsonObject.number(key: String) = this[key]?.jsonPrimitive?.doubleOrNull

    /**
     * The message type travels in the SSE `event:` name, not in the payload —
     * the server strips it (`const { type, ...data } = message`) before
     * writing the frame. Keying on the payload instead decoded nothing at all,
     * which is exactly how this was found.
     */
    fun message(event: String?, raw: String): BrowserLiveMessage? {
        if (event == null) return null
        val root = runCatching { json.parseToJsonElement(raw).jsonObject }.getOrNull()
            ?: JsonObject(emptyMap())

        return when (event) {
            "frame" -> {
                val seq = root["seq"]?.jsonPrimitive?.intOrNull ?: return null
                val data = root.string("data") ?: return null
                val metadata = runCatching { root["metadata"]!!.jsonObject }.getOrNull() ?: return null
                val width = metadata.number("deviceWidth") ?: return null
                val height = metadata.number("deviceHeight") ?: return null
                BrowserLiveMessage.Frame(
                    BrowserFrame(seq, data, root.string("format") ?: "jpeg", width, height),
                )
            }

            "status" -> BrowserLiveMessage.Status(
                BrowserStatus(
                    connected = root.bool("connected") ?: return null,
                    screencasting = root.bool("screencasting") ?: false,
                    viewportWidth = root.number("viewportWidth") ?: 1280.0,
                    viewportHeight = root.number("viewportHeight") ?: 720.0,
                ),
            )

            "url" -> BrowserLiveMessage.Url(root.string("url") ?: return null)

            "tabs" -> {
                val array = runCatching { root["tabs"]!!.jsonArray }.getOrNull() ?: return null
                BrowserLiveMessage.Tabs(
                    array.mapNotNull { element ->
                        val tab = runCatching { element.jsonObject }.getOrNull() ?: return@mapNotNull null
                        val id = tab.string("tabId") ?: return@mapNotNull null
                        BrowserTab(
                            tabId = id,
                            title = tab.string("title") ?: "",
                            url = tab.string("url") ?: "",
                            active = tab.bool("active") ?: false,
                        )
                    },
                )
            }

            // The server names this `ready`, and it is the only place a viewer
            // id ever arrives. Without it there is nothing to post against.
            "ready" -> BrowserLiveMessage.Ready(root.string("viewerId") ?: return null)

            "control" -> BrowserLiveMessage.Control(
                controlling = root.bool("controlling") ?: false,
                held = root.bool("held") ?: false,
            )

            "heartbeat" -> BrowserLiveMessage.Heartbeat

            "error" -> BrowserLiveMessage.Error(
                root.string("message") ?: "The browser stream was interrupted.",
            )

            // A name we do not know is a protocol change, not a reason to tear
            // down a stream the person is watching.
            else -> null
        }
    }
}

/** Decoded frame bytes, or null when the base64 was not what it claimed.
 *
 * java.util.Base64 rather than android.util.Base64: :core is a plain JVM
 * module so its tests run without an emulator, and pulling in an Android type
 * here would end that. */
fun BrowserFrame.bytes(): ByteArray? =
    runCatching { java.util.Base64.getDecoder().decode(data) }.getOrNull()

/**
 * The browser-live transport: one SSE stream in, one action channel out.
 *
 * Mirrors CompanionCore's `BrowserLiveClient`.
 */
class BrowserLiveTransport internal constructor(
    private val base: HttpUrl,
    private val token: String?,
    private val streamClient: OkHttpClient,
    private val actionClient: OkHttpClient,
) {
    private val jsonMedia = "application/json".toMediaType()

    private fun request(path: String): Request.Builder =
        Request.Builder()
            .url(base.newBuilder().encodedPath(path).build())
            .apply { token?.let { header("Authorization", "Bearer $it") } }

    /**
     * The frame stream. Runs until the server ends it or the collector leaves;
     * reconnection belongs to whatever knows if the view is still on screen,
     * exactly as it does for the main event stream.
     */
    fun live(botId: String): Flow<BrowserLiveMessage> = callbackFlow {
        val call = streamClient.newCall(request("/api/bots/$botId/browser/live").get().build())
        val responseRef = AtomicReference<Response?>(null)
        val reader = launch(Dispatchers.IO) {
            try {
                val response = call.execute()
                responseRef.set(response)
                // Read the refusal rather than discard it. Throwing a bare
                // status made every 403 read as "browser control is off",
                // whatever the server actually said.
                if (response.code != 200) {
                    throw APIError.Status(response.code, browserErrorText(response.peekBody(4096).string()))
                }
                val body = response.body ?: throw APIError.Transport("The browser sent an empty stream.")
                val source = body.source()
                val parser = SSEParser()

                while (isActive) {
                    val line = source.readUtf8Line() ?: break
                    val event = parser.line(line) ?: continue
                    val message = BrowserLiveDecoder.message(event.event, event.data) ?: continue
                    send(message)
                }
                parser.reset()
                close()
            } catch (_: CancellationException) {
                // A collector leaving the flow deliberately tears down the call.
            } catch (error: APIError) {
                close(error)
            } catch (error: IOException) {
                close(APIError.Transport(error.message ?: "Could not reach the browser.", error))
            } finally {
                responseRef.getAndSet(null)?.close()
            }
        }

        awaitClose {
            responseRef.getAndSet(null)?.close()
            call.cancel()
            reader.cancel()
        }
    }

    /**
     * One action. The server answers only after the browser has applied it,
     * which is what makes the queue's one-in-flight rule necessary.
     */
    suspend fun action(botId: String, viewerId: String, body: JsonObject) = withContext(Dispatchers.IO) {
        val payload = buildJsonObject {
            body.forEach { (key, value) -> put(key, value) }
            put("viewerId", viewerId)
        }
        val request = request("/api/bots/$botId/browser/action")
            .post(payload.toString().toRequestBody(jsonMedia))
            .build()
        actionClient.newCall(request).execute().use { response ->
            if (response.code != 200) {
                throw APIError.Status(response.code, browserErrorText(response.peekBody(4096).string()))
            }
        }
    }

    /** An input body, posted through the same channel. */
    suspend fun send(botId: String, viewerId: String, input: BrowserInputBody) {
        val encoded = Json.encodeToJsonElement(BrowserInputBody.serializer(), input).jsonObject
        action(botId, viewerId, encoded)
    }
}

/** The `error` field of a JSON refusal, which is how both the sidecar and the
 * harness explain themselves. Null for anything else. */
fun browserErrorText(body: String): String? = runCatching {
    Json.parseToJsonElement(body).jsonObject["error"]?.jsonPrimitive?.contentOrNull
}.getOrNull()
